pkg/southbound/synchronizer/device_update.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package synchronizer
+
+import (
+	"errors"
+	"strconv"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cenkalti/backoff"
+
+	"github.com/onosproject/onos-config/pkg/events"
+
+	topodevice "github.com/onosproject/onos-topo/api/device"
+)
+
+func (s *Session) getTermPerDevice(device *topodevice.Device) (int, error) {
+	term := device.Attributes[mastershipTermKey]
+	if term == "" {
+		return 0, nil
+	}
+	return strconv.Atoi(term)
+
+}
+
+func (s *Session) updateDevice(connectivity topodevice.ConnectivityState, channel topodevice.ChannelState,
+	service topodevice.ServiceState) error {
+	log.Info("Update device state")
+
+	id := s.device.ID
+	topoDevice, err := s.deviceStore.Get(id)
+	st, ok := status.FromError(err)
+
+	// If the device doesn't exist then we should not update its state
+	if ok && err != nil && st.Code() == codes.NotFound {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	protocolState, index := containsGnmi(topoDevice.Protocols)
+	if protocolState != nil {
+		topoDevice.Protocols = remove(topoDevice.Protocols, index)
+	} else {
+		protocolState = new(topodevice.ProtocolState)
+	}
+
+	protocolState.Protocol = topodevice.Protocol_GNMI
+	protocolState.ConnectivityState = connectivity
+	protocolState.ChannelState = channel
+	protocolState.ServiceState = service
+	topoDevice.Protocols = append(topoDevice.Protocols, protocolState)
+	mastershipState, err := s.mastershipStore.GetMastership(topoDevice.ID)
```

Comment:
This mastership state doesn’t necessarily represent the state on which this update should be based. Since we’re reading from the mastership store on every update attempt, it’s possible for this node to read a _different_ (newer) mastership state (e.g. for a new master/term) than this update is based on. Who knows if this node is even the master any more. Rather than reading from the store here, this function should take as an argument the mastership state on which the update is based so that every attempt to perform the update is based on that state. Reading stores multiple times can alway introduce this sort of consistency issue. The only reason we should read the store here is to determine whether the state on which the update is based should be cancelled due to a new master/higher term.